src/scenic/core/propositions.py

```diff
@@ -0,0 +1,199 @@
+"""Objects representing propositions that can be used to specify conditions"""
+
+from functools import reduce
+import operator
+import rv_ltl
+from scenic.core.errors import InvalidScenarioError
+
+from scenic.core.lazy_eval import needsLazyEvaluation
+
+class PropositionMonitor:
+	def __init__(self, proposition: "PropositionNode") -> None:
+		self._proposition = proposition
+		self._monitor = proposition.ltl_node.create_monitor()
+		
+	def update(self):
+		atomic_propositions = self._proposition.atomics()
+		state = {}
+		for ap in atomic_propositions:
+			b = ap.closure()
+			if needsLazyEvaluation(b):
+				raise InvalidScenarioError(f'value undefined outside of object definition')
+			state[str(ap.syntax_id)] = b
+		self._monitor.update(state)
+		return self._monitor.evaluate()
+
+
+class PropositionNode:
+	"""Base class for temporal and non-temporal propositions"""
+	def __init__(self, syntax_id, ltl_node) -> None:
+		self.syntax_id = syntax_id
+		self.ltl_node = ltl_node
+
+		self.is_temporal = False
+		"""tells if the proposition is temporal"""
+
+	def check_constrains_sampling(self):
+		"""checks if the proposition can be used for pruning
+
+		A requirement can be used for pruning if it is evaluated on the scene generation phase before simulation, and
+		violation in that phase immediately results in discarding the scene and regenerating a new one.
+
+		For simplicity, we currently check two special cases:
+		1. requirements with no temporal requirement
+		2. requirements with only one `always` operator on top-level
+
+		Returns:
+			None|int: returns syntax id that can be used for pruning, None if proposition cannot be used for pruning
+		"""
+		node = self
+
+		# if `always` is on top-level, check what's inside
+		if isinstance(node, Always):
+			node = node.req
+
+		eligible = True
+		for n in node.flatten():
+			# turns false if any one of the node is temporal
+			eligible = eligible and (not n.is_temporal)
```

Comment:
I find this a little awkward to read; what about `eligible = all(not n.is_temporal for n in node.flatten())`?

Reply:
Good point. Fixed in 1f01cda92fbea858e0fbb053461a2e49b0231796